Constrain user input queries by domain

millerdev opened the pull request:
Small optimization. Previously the CTE query filtered on user input was doing a sequential scan on the full table. Now it can use the domain to eliminated more rows from the root nodes of the recursive query.

This removed one of two sequential scans. I don't understand yet why it's doing the second sequential scan. More research is required.

## Safety Assurance

### Safety story

This should not change the meaning of the query since all selected locations must belong to the domain.

### Automated test coverage

Yes.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

millerdev committed:
Constrain user input queries by domain

Previously the inner CTE query would do a sequential scan on the full
table. Now it can use the domain index.

gherceg commented on corehq/apps/locations/models.py:
```diff
@@ -331,6 +331,7 @@ def filter_by_user_input(self, domain, user_input):
                     query = query.filter(
                         Q(name__icontains=part) | Q(site_code__icontains=part)
                     )
+                query = query.filter(domain=domain)
```
Is my understanding right that this didn't restrict by domain previously, let alone use domain in this function? Was it relying on other filters added elsewhere to do that?

millerdev replied:
Yes. Usually the `accessible_to_user` filter is also used in conjunction with this one. [`CallCenterOptionsController.get_locations_query`](https://github.com/dimagi/commcare-hq/blob/62febbd8187c27bd5eedfc764d924b59426534e9/corehq/apps/callcenter/views.py#L74-L77) is the exception; I'm not sure how that one is/was being filtered by domain.

esoergel replied:
Looks like this used to filter by domain (which is why it's an arg), but I removed that in https://github.com/dimagi/commcare-hq/pull/26671 - must've been an oversight

gherceg commented on corehq/apps/locations/tests/test_location_queries.py:
```diff
@@ -149,13 +149,12 @@ def test_location_restricted_but_unassigned(self):
         self.restrict_user_to_assigned_locations(self.web_user)
 
+        mass_locs = SQLLocation.objects.filter_by_user_input(self.domain, "Massachusetts/")
+        self.assertIn(f'"domain" = {self.domain})', str(mass_locs.query))
```
It took me a second to realize what exactly this was doing. I imagine a more "traditional" test that creates two locations in different domains with the same name, and asserts that only the location from the passed in domain is returned would be more intuitive? Just a personal preference though.

millerdev replied:
Good point. https://github.com/dimagi/commcare-hq/pull/36216/commits/b938ff4d0c0de53668edc7368ef6593f8b96491a

millerdev committed:
Refactor: test case per path query

millerdev committed:
Improve filter_by_user_input test coverage

This test fails if `query = query.filter(domain=domain)` is commented
out in `LocationManager.filter_by_user_input` (added two commits prior).

Also revert change to test_filter_by_user_input, which wasn't a great
way to test the domain filter.

gherceg commented on corehq/apps/locations/tests/test_location_queries.py:
```diff
@@ -191,30 +192,39 @@ class TestFilterByUserInput(LocationHierarchyTestCase):
+    ('Middl/camb', ['Cambridge', 'Cambridge']),
+], TestFilterByUserInput)
+def test_path_query(self, querystring, expected):
+    self.assert_query_has_results(querystring, expected)
```
Not at all blocking, but just curious to hear more about the rationale for putting assertions in helpers. I like keeping asserts simple, even if that means have more code/duplicate code since it is easy to see what is being asserted in the test, rather than having to go find the helper. That is obviously a nit, and more of an issue when reviewing PRs than working with the tests.

millerdev replied:
The thought of duplicating the logic in that assertion in each of the new test cases seems worse to me than putting it in a concise assert method.

Previously it was the body of a for-loop where each loop was effectively a separate test case, except not actually. Do you think that was better?

gherceg replied:
Yeah sorry I wasn't clear about which duplication I was referring to. You could just have the logic that is in `assert_query_has_results` in `test_path_query` directly. Sure, there might be other tests that do a similar assertion, but that is the tradeoff with duplication that I'm ok with. Recognize that this is just a personal preference though. I am a fan of using `generate_cases` 👍🏻

millerdev replied:
We were on the same page with regard to what you were suggesting, but I think we disagree on the best way to write it. I don't like the idea of duplicating the body of `assert_query_has_results` in multiple tests; I like the helper function.

gherceg replied:
Yeah so I think we were not on the same page. The helper assert is what I was commenting on in the first place 😅 with the main rationale being it makes reviewing changes to tests a little unclear because I have to go find the helper to understand what is being asserted (take the last commit in this PR for example). But again, happy to agree to disagree!

esoergel replied:
For what it's worth I'm on team "minimize duplication in tests" in general too.  Some coding book I read years ago preached that tests merit the same considerations you'd give to production code, which resonates with me.  I've worked in test suites with a lot of duplication before and found it hard to make cross-cutting updates.  At the extremes they are verbose and hard to read (like if each case here was its own test without a helper, which I know no one is advocating for, but which we do elsewhere).

That said, [Locality of Behavior](https://htmx.org/essays/locality-of-behaviour/) is a super important consideration for tests, though that's not violated if the abstractions are clean and intuitive.

gherceg reviewed (approved):
Nice!
